### PR TITLE
[imgcodec] Add thread index and cuda stream to Decode APIs

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature** 

## Description:

- It adds a thread_idx to BatchParallelDecoderImpl Decode API, so that decoders can have per-thread resources and refer to them via thread_idx.
- It adds a cuda stream to all GPU Decode APIs, so that the work can be either scheduled on that stream or we can synchronize with it

### Affected modules and functionalities:

decoder_parallel_impl.h

### Key points relevant for the review:

NA

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/decoders/decoder_parallel_impl.h [hunk 1]
```diff
@@ -74,13 +97,69 @@ class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
...
+   * @param stream CUDA stream to synchronize with
+   * @param thread_idx thread index in the thread pool
+   * @param out output sample view
+   * @param in encoded image source
+   * @param opts decoding parameters
+   * @param roi region-of-interest
+   * @return std::vector<DecodeResult> 
+   */
+  virtual DecodeResult DecodeImplTask(cudaStream_t stream,
+                                      int thread_idx,
```
Comment (mzient) [inline dali/imgcodec/decoders/decoder_parallel_impl.h:156]: Perhaps here thread_idx should go first - the thread_idx is something added on top of the existing parameters.

**Review by mzient — APPROVED** [review]
OK, with one minor issue.

File: dali/imgcodec/decoders/decoder_parallel_impl.h (hunk 1 above)
Comment (jantonguirao) [reply dali/imgcodec/decoders/decoder_parallel_impl.h:156]: Agree. I'll fix

**Review by mzient — APPROVED** [review]

File: dali/imgcodec/decoders/decoder_parallel_impl.h [hunk 2]
```diff
@@ -74,13 +97,69 @@ class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
...
+                                               cspan<ImageSource *> in,
+                                               DecodeParams opts,
+                                               cspan<ROI> rois) {
+    assert(out.size() == in.size());
+    assert(rois.empty() || rois.size() == in.size());
+    std::vector<DecodeResult> ret(out.size());
+    ROI no_roi;
+    for (int i = 0; i < in.size(); i++) {
+      tp_->AddWork([&, i](int tid) {
+        ret[i] = DecodeImplTask(stream, tid, out[i], in[i], opts, rois.empty() ? no_roi : rois[i]);
```
Comment (staniewzki) [inline dali/imgcodec/decoders/decoder_parallel_impl.h:119]:
```suggestion
        ret[i] = DecodeImplTask(tid, stream, out[i], in[i], opts, rois.empty() ? no_roi : rois[i]);
```
The order of arguments is wrong.

File: dali/imgcodec/decoders/decoder_parallel_impl.h [hunk 3]
```diff
@@ -57,15 +57,38 @@ class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
...
+  }
+
+  DecodeResult Decode(cudaStream_t stream,
+                      SampleView<GPUBackend> out,
+                      ImageSource *in,
+                      DecodeParams opts,
+                      const ROI &roi) override {
+    DecodeResult ret;
+    tp_->AddWork([&](int tid) {
+      ret = DecodeImplTask(stream, tid, out, in, opts, roi);
```
Comment (staniewzki) [inline dali/imgcodec/decoders/decoder_parallel_impl.h:85]:
```suggestion
      ret = DecodeImplTask(tid, stream, out, in, opts, roi);
```

File: dali/imgcodec/decoders/decoder_parallel_impl.h (hunk 2 above)
Comment (jantonguirao) [reply dali/imgcodec/decoders/decoder_parallel_impl.h:119]: true. Thanks for spotting that

File: dali/imgcodec/decoders/decoder_impl.h [hunk 4]
```diff
@@ -64,21 +64,20 @@ class DLL_PUBLIC ImageDecoderImpl : public ImageDecoderInstance {
   /**
    * @brief To be overriden by a GPU/mixed codec implementation
    */
-  DecodeResult Decode(SampleView<GPUBackend> out, ImageSource *in,
+  DecodeResult Decode(cudaStream_t stream, SampleView<GPUBackend> out, ImageSource *in,
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/imgcodec/decoders/decoder_impl.h:67]: It was actually the original proposal, but we discussed it and we agreed that executing on default stream by accident is a great source of hidden bugs and therefore it'd be better not to allow a default value there. Also, by putting the cuda stream at the front we are able to pack all the other arguments as in the CPU counterparts.

**Review by szalpal — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**